base/docker-integration.go

```diff
@@ -0,0 +1,313 @@
+			{"bucket_admin", DefaultTestIndexBucketname},
+		},
+	}
+
+	//Connect to cluster with gocb and create users and buckets
+	cluster, err = gocb.Connect("http://localhost:8091/")
+	fatalError(`Unable to connect to cluster with gocb`, err)
+	manager := cluster.Manager(DefaultCouchbaseAdministrator, DefaultCouchbasePassword)
+
+	err = manager.UpdateBucket(&bucketSettings)
+	fatalError(`Unable to create bucket`, err)
+	err = manager.UpdateBucket(&bucketSettings1)
+	fatalError(`Unable to create bucket`, err)
+	err = manager.UpdateBucket(&bucketSettings2)
+	fatalError(`Unable to create bucket`, err)
+
+	err = manager.UpsertUser("local", DefaultTestUsername, &userSettings)
+	fatalError(`Unable to create user`, err)
+	err = manager.UpsertUser("local", DefaultTestShadowUsername, &userSettings1)
+	fatalError(`Unable to create user`, err)
+	err = manager.UpsertUser("local", DefaultTestIndexUsername, &userSettings2)
+	fatalError(`Unable to create user`, err)
+
+	err = cluster.Close()
+	if err != nil {
+		fmt.Println("Failed to close gocb cluster connection but will continue")
+	}
+
+	areReady := func(url string, checkReady func(reader io.ReadCloser) bool) {
+		for {
+			req, err := http.NewRequest("GET", url, nil)
+			fatalError(`Failed to poll status`, err)
+			req.SetBasicAuth(DefaultCouchbaseAdministrator, DefaultCouchbasePassword)
+			response, err = client.Do(req)
+			fatalError(`Failed to poll status`, err)
+			if checkReady(response.Body) {
+				return
+			}
+		}
+	}
+
+	//Poll buckets to ensure they are all healthy and ready to work
+	bucketReadyCheck := func(reader io.ReadCloser) bool {
+		var jsonBody []responseBody
+		err = json.NewDecoder(reader).Decode(&jsonBody)
+		fatalError(`Failed to poll status`, err)
+		return areBucketsReady(&jsonBody)
+	}
+	areReady("http://localhost:8091/pools/default/buckets", bucketReadyCheck)
+
+	//Poll nodes / node to ensure they are all healthy and ready to work
+	nodeReadyCheck := func(reader io.ReadCloser) bool {
+		var jsonBody responseBody
+		err = json.NewDecoder(reader).Decode(&jsonBody)
+		fatalError(`Failed to poll status`, err)
+		return isNodeReady(&jsonBody)
+	}
+	areReady("http://localhost:8091/pools/default", nodeReadyCheck)
+
+	//Run integration tests
+	code := t.Run()
+
+	//Remove Docker container when complete
+	if err := pool.Purge(resource); err != nil {
+		log.Fatalf("Could not purge resource: %s", err)
+	}
+
+	os.Exit(code)
+}
+
+func areBucketsReady(res *[]responseBody) bool {
+	allReady := true
+	nres := *res
+	for i := 0; i < len(nres); i++ {
+		if len(nres[i].Nodes) == 1 && nres[i].Nodes[0].Status != "healthy" {
+			return false
+		}
+	}
+	return allReady
+}
+
+func isNodeReady(res *responseBody) bool {
+	ready := true
+	nres := *res
+	//TODO: Figure out a nice way of handling this other than waiting for 60 secs of uptime
+	uptime, err := strconv.Atoi(nres.Nodes[0].Uptime)
+	if err != nil || len(nres.Nodes) == 1 && nres.Nodes[0].Status != "healthy" || uptime <= 60 {
```

Comment:
This uptime check is pretty painful to wait for, and actually the selection of tests I just ran worked without waiting for it. Can we determine _what_ breaks here?

Reply:
From my tests the thing that breaks appears to be N1QL. Gets the no nodes available error. Sometimes it works after 10 seconds. Sometimes takes a minute, I guess it depends on environmental factors.

Reply:
OK. Well there's a TODO in the code, so I think we can park it like this if it's non-trivial.